api/clients/v2/eigenda_client.go

```diff
@@ -0,0 +1,279 @@
+		return nil, fmt.Errorf("verify cert with timeout for blobKey %v: %w", blobKey, err)
+	}
+
+	relayKeys := eigenDACert.BlobVerificationProof.BlobCertificate.RelayKeys
+	relayKeyCount := len(relayKeys)
+	if relayKeyCount == 0 {
+		return nil, errors.New("relay key count is zero")
+	}
+
+	blobCommitmentProto := contractEigenDABlobVerifier.BlobCommitmentBindingToProto(
+		&eigenDACert.BlobVerificationProof.BlobCertificate.BlobHeader.Commitment)
+	blobCommitment, err := encoding.BlobCommitmentsFromProtobuf(blobCommitmentProto)
+
+	if err != nil {
+		return nil, fmt.Errorf("blob commitments from protobuf: %w", err)
+	}
+
+	// create a randomized array of indices, so that it isn't always the first relay in the list which gets hit
+	indices := c.random.Perm(relayKeyCount)
+
+	// TODO (litt3): consider creating a utility which can deprioritize relays that fail to respond (or respond maliciously)
+
+	// iterate over relays in random order, until we are able to get the blob from someone
+	for _, val := range indices {
+		relayKey := relayKeys[val]
+
+		blob, err := c.getBlobWithTimeout(ctx, relayKey, blobKey)
+		// if GetBlob returned an error, try calling a different relay
+		if err != nil {
+			c.log.Warn("blob couldn't be retrieved from relay", "blobKey", blobKey, "relayKey", relayKey, "error", err)
+			continue
+		}
+
+		err = c.verifyBlobAgainstCert(blobKey, relayKey, blob, blobCommitment.Commitment, blobCommitment.Length)
+
+		// An honest relay should never send a blob which doesn't verify
+		if err != nil {
+			c.log.Warn("verify blob from relay: %w", err)
+			continue
+		}
+
+		payload, err := c.codec.DecodeBlob(blob)
+		if err != nil {
+			c.log.Error(
+				`Blob verification was successful, but decode blob failed!
+					This is likely a problem with the local blob codec configuration,
+					but could potentially indicate a maliciously generated blob certificate.
+					It should not be possible for an honestly generated certificate to verify
+					for an invalid blob!`,
+				"blobKey", blobKey, "relayKey", relayKey, "eigenDACert", eigenDACert, "error", err)
+			return nil, fmt.Errorf("decode blob: %w", err)
+		}
+
+		return payload, nil
+	}
+
+	return nil, fmt.Errorf("unable to retrieve blob %v from any relay. relay count: %d", blobKey, relayKeyCount)
+}
+
+// verifyBlobAgainstCert verifies the blob received from a relay against the certificate.
+// This method does NOT verify the blob with an eth_call to verifyBlobV2, that must be done separately.
+//
+// The following verifications are performed in this method:
+// 1. Verify that blob isn't empty
+// 2. Verify the blob kzg commitment
+// 3. Verify that the blob length is less than or equal to the claimed blob length
+//
+// If all verifications succeed, the method returns nil. Otherwise, it returns an error.
+func (c *EigenDAClient) verifyBlobAgainstCert(
+	blobKey core.BlobKey,
+	relayKey core.RelayKey,
+	blob []byte,
+	kzgCommitment *encoding.G1Commitment,
+	blobLength uint) error {
```

Comment:
shouldn't this method take a blob and a cert? That's what I would have intuited from the name. Is it really worth extracting all of these fields from the cert? Would we ever want to call it without actually having a cert in hand?

Reply:
The problem with passing in the entire `EigenDACert` is that `verifyBlobAgainstCert` is called in a loop, and the data contained in the cert must be processed before it can be used (i.e., binding type must be converted to the internal type). And we wouldn't want to do this conversion multiple times.

One option would be to pass in an `encoding.BlobCommitments` type, in place of the separate `G1Commitment` and `uint` length. But since these 2 fields are only a subset of the data contained in the `encoding.BlobCommitments` struct, my personal preference is to only pass in the things we need.

Reply:
Gotcha that makes sense at least. Might be nice to have an internal type for Cert (contains all the datatypes that a cert has, but in the internal types instead of contract types). But I'm fine with merging without this. This PR has been alive for too long already haha.

Reply:
I think we need a larger team discussion about when to create and use internal datatypes.

Should we always have internal types? Only when we need to define methods for autogenerated types? Only when necessary (i.e., use proto types wherever possible)?